src/python_testing/mdns_discovery/mdns_discovery.py

```diff
@@ -0,0 +1,274 @@
+            expected response time for service announcements.
+
+        Returns:
+            None: This method does not return any value. The results of the service discovery
+                  are handled by the callback function.
+        """
+        self._zc = Zeroconf(ip_version=IPVersion.V6Only)
+        self._discovered_services = {mdns_type.name: [] for mdns_type in MdnsType}
+        self._aiobrowser = AsyncServiceBrowser(zeroconf=self._zc,
+                                               type_=self._service_types,
+                                               handlers=[self._on_service_state_change]
+                                               )
+        # Wait for discovery
+        await asyncio.sleep(discovery_duration_sec)
+        self._discovery_performed = True
+        await self._async_close()
+
+    def _on_service_state_change(
+        self,
+        zeroconf: Zeroconf,
+        service_type: str,
+        name: str,
+        state_change: ServiceStateChange
+    ) -> None:
+        """
+        Callback method triggered on mDNS service state change.
+
+        This method is called by the Zeroconf library when there is a change in the state of an mDNS service.
+        It handles the addition of new services by initiating a query for their detailed information.
+
+        Args:
+            zeroconf (Zeroconf): The Zeroconf instance managing the network operations.
+            service_type (str): The type of the mDNS service that changed state.
+            name (str): The name of the mDNS service.
+            state_change (ServiceStateChange): The type of state change that occurred.
+
+        Returns:
+            None: This method does not return any value.
+        """
+        if state_change.value == ServiceStateChange.Added.value:
+            asyncio.ensure_future(self._query_service_info(
+                zeroconf,
+                service_type,
+                name)
+            )
+
+    async def _query_service_info(self, zeroconf: Zeroconf, service_type: str, name: str) -> None:
+        """
+        This method requests detailed information about an mDNS service, such as its address,
+        port, and other service metadata. The information is requested asynchronously.
+
+        Args:
+            zeroconf (Zeroconf): The Zeroconf instance managing the network operations.
+            service_type (str): The type of the mDNS service to query.
+            name (str): The name of the mDNS service.
+
+        Returns:
+            None: This method does not return any value. The results of the query are processed internally.
+        """
+        # Get service info
+        service_info = AsyncServiceInfo(service_type, name)
+        is_service_discovered = await service_info.async_request(zeroconf, 3000)
+
+        if is_service_discovered:
+            # Add service info to discovered services
+            for mdns_type in MdnsType:
+                if service_type == mdns_type.value:
```

Comment:
if you query for a specific operational node, will this service name match? ie, if you query for `<fabricid>-<nodeid>._matter._tcp.local`, does the incoming service_type variable include the `<fabricid>-<nodeid>` portion of the name? 

Also, it seems like it would make more sense to build this dictionary based on the queries, rather than keeping slots open for each of the types. Then queries for subtypes can be differentiated.

Reply:
Resolved